### ISSUE TITLE
File encoding should be detected correctly

### ISSUE THREAD

**Issue by The-Compiler** [PyCQA/pydocstyle#70]
pep257 fails to open a file with an unicode arrow:

```
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 17236: character maps to <undefined>
```

This is because Windows uses the cp1252 encoding by default, but my source file is encoded as utf-8.

Instead of just calling `open()` on the file, it should detect the correct encoding according to [PEP263](http://legacy.python.org/dev/peps/pep-0263/) and open it with that encoding.

**Comment by jquast** [comment]
Here is a PR for another tool implementing precisely that, https://github.com/landscapeio/prospector/pull/71

**Comment by The-Compiler** [comment]
There's [tokenize.detect_encoding](https://docs.python.org/3/library/tokenize.html#tokenize.detect_encoding) and in later Python versions, [tokenize.open](https://docs.python.org/3/library/tokenize.html#tokenize.open) which is probably a lot simpler than that PR :wink:

**Comment by jquast** [comment]
nice hint @The-Compiler, unfortunately you're right, it isn't available in python2.7, presume pep257 aims support for python2.7

**Comment by peterjc** [comment]
I just traced a problem in my plugin based on ``pydocstyle`` to this bit of code:

https://github.com/PyCQA/pydocstyle/blob/2.0.0/src/pydocstyle/checker.py#L23

```python
# If possible (python >= 3.2) use tokenize.open to open files, so PEP 263
# encoding markers are interpreted.
try:
    tokenize_open = tk.open
except AttributeError:
    tokenize_open = open
```

Would a backport of ``tokenize.open`` for Python 2 be acceptable here?

Note it seems that because later the contents of the file get executed, it is helpful to also remove/replace the encoding line to avoid triggering ``SyntaxError: encoding declaration in Unicode string`` when passing it to ``exec(...)``:

https://github.com/PyCQA/pydocstyle/blob/2.0.0/src/pydocstyle/parser.py#L286

```python
        try:
            compile(src, filename, 'exec')
        except SyntaxError as error:
            six.raise_from(ParseError(), error)
```

I'm not proud of this hacked together back port, but it might serve if Python 2.7 support is desired:

https://github.com/peterjc/flake8-rst-docstrings/commit/80697a645c1d570a9c8b9fad226f5c7078d30a98

**Comment by peterjc** [comment]
See also pull request #152 which added a test case using ``# -*- coding: Windows-1255 -*``

**Comment by sigmavirus24** [comment]
pycodestyle has code to work around the absence of `tokenize.open`. Flake8 has taken and improved it as well.

**Comment by peterjc** [comment]
@sigmavirus24 Can you point me at this code please? And can this issue be closed?

**Comment by aphedges** [comment]
pydocstyle has been using `tokenizer.open` when reading source files since 4948f22d962d422733b1edadc0852a53444d8a37. This is therefore no longer a problem.
